### PR TITLE
Improve aos_to_soa3/4 and soa_to_aos3/4 for 8 and 16 wide targets.

### PR DESCRIPTION
**Pull request by dbabokin**
Performance improvements for `aos_to_soa3()`, `aos_to_soa4()`, `soa_to_aos3()`, and `soa_to_aos4()` for 8-wide and 16-wide CPU targets:
- all aos_to_soa3/4 and soa_to_aos3/4 implementations (with 2 exceptions below) were changed to use "wide" shuffles instead of 2- or 4-pumped implementation reducing the algorithm to 4-wide shuffles.
- `aos_to_soa3()` for `double`/`int64` types has different algorithm, which is faster on medium-size data sets.
- `soa_to_aos4()` for `double`/`int64` types was not changed for AVX512 targets, as new algorithm is substantially slower (16%-28%) on these targets.

Along with performance improvements there are other related changes:
- added build dependency for .ll target files on the files they include.
- added `ISA` macro to CPU target files, so it's possible to differentiate on target ISA in `util.m4`, same macro is used in `svml()` call now.
- removed executable permission for `builtins-c-common.hpp` file.
- benchmarks report ISPC target they are built for and ISPC switches used to build them.
- added a "warmup" workload for benchmarks, which is the first in the list now. This is not necessary on well stabilized machines, but helps getting more stable results when the machine is unstable for some reason.
- benchmarks README updated with recommendations how to have more stable runs.

### IceLake Server machine (AVX512): `avx512skx-i32x8` and `avx512skx-i32x16`
|   |ref-x8 (ns)|new-x8 (ns)|%|ref-x16 (ns)|new-x16 (ns)|%|
|---|---:|---:|---:|---:|---:|---:|
|aos_to_soa4_stdlib_float/256|36.7|18.4|49.86%|45.8|10.3|77.51%|
|aos_to_soa4_stdlib_float/4096|587|293|50.09%|734|188|74.39%|
|aos_to_soa4_stdlib_float/32768|5063|2515|50.33%|5869|2972|49.36%|
|aos_to_soa4_stdlib_float/1048576|300311|282810|5.83%|291727|289534|0.75%|
|aos_to_soa4_stdlib_double/256|36.7|45.9|-25.07%|36.7|28.3|22.89%|
|aos_to_soa4_stdlib_double/4096|982|734|25.25%|982|745|24.13%|
|aos_to_soa4_stdlib_double/32768|7857|5946|24.32%|7875|5956|24.37%|
|aos_to_soa4_stdlib_double/1048576|594639|566381|4.75%|604695|582501|3.67%|
|aos_to_soa3_stdlib_float/192|14.7|11.5|21.77%|15.8|8.13|48.54%|
|aos_to_soa3_stdlib_float/3072|227|183|19.38%|239|144|39.75%|
|aos_to_soa3_stdlib_float/24576|4378|1901|56.58%|4429|2329|47.41%|
|aos_to_soa3_stdlib_float/786432|232036|211743|8.75%|222782|216294|2.91%|
|aos_to_soa3_stdlib_double/192|18.8|18.6|1.06%|18.4|17.3|5.98%|
|aos_to_soa3_stdlib_double/3072|314|314|0.00%|339|251|25.96%|
|aos_to_soa3_stdlib_double/24576|6639|4339|34.64%|6639|4418|33.45%|
|aos_to_soa3_stdlib_double/786432|454314|424178|6.63%|466232|435335|6.63%|
||||||||
|soa_to_aos4_stdlib_float/256|36.7|18.3|50.14%|36.7|12.6|65.67%|
|soa_to_aos4_stdlib_float/4096|587|296|49.57%|587|184|68.65%|
|soa_to_aos4_stdlib_float/32768|4716|2514|46.69%|4696|3380|28.02%|
|soa_to_aos4_stdlib_float/1048576|283522|282537|0.35%|283466|290793|-2.58%|
|soa_to_aos4_stdlib_double/256|36.7|same|0%|36.7|same|0%|
|soa_to_aos4_stdlib_double/4096|630|same|0%|630|same|0%|
|soa_to_aos4_stdlib_double/32768|5005|same|0%|5006|same|0%|
|soa_to_aos4_stdlib_double/1048576|564914|same|0%|566631|same|0%|
|soa_to_aos3_stdlib_float/192|18.3|13.8|24.59%|18.3|8.99|50.87%|
|soa_to_aos3_stdlib_float/3072|543|220|59.48%|293|128|56.31%|
|soa_to_aos3_stdlib_float/24576|2840|1953|31.23%|2364|2248|4.91%|
|soa_to_aos3_stdlib_float/786432|212294|212160|0.06%|211985|216387|-2.08%|
|soa_to_aos3_stdlib_double/192|27.5|20.6|25.09%|27.5|17.4|36.73%|
|soa_to_aos3_stdlib_double/3072|441|330|25.17%|440|280|36.36%|
|soa_to_aos3_stdlib_double/24576|4003|4303|-7.49%|3915|4414|-12.75%|
|soa_to_aos3_stdlib_double/786432|423992|424327|-0.08%|424208|434241|-2.37%|

### CoffeeLake (Intel® Core™ i9-9900K CPU @ 3.60GHz) (AVX2): `avx2-i32x8` and `avx2-i32x16`
|   |ref-x8 (ns)|new-x8 (ns)|%|ref-x16 (ns)|new-x16 (ns)|%|
|---|---:|---:|---:|---:|---:|---:|
|aos_to_soa4_stdlib_float/256|26.2|19.7|24.81%|26.3|27.9|-6.08%|
|aos_to_soa4_stdlib_float/4096|423|318|24.82%|420|450|-7.14%|
|aos_to_soa4_stdlib_float/32768|4894|3390|30.73%|4618|4674|-1.21%|
|aos_to_soa4_stdlib_float/1048576|189279|155925|17.62%|184063|187405|-1.82%|
|aos_to_soa4_stdlib_double/256|46|32.8|28.70%|46|32.8|28.70%|
|aos_to_soa4_stdlib_double/4096|738|577|21.82%|735|837|-13.88%|
|aos_to_soa4_stdlib_double/32768|9395|8901|5.26%|9920|10035|-1.16%|
|aos_to_soa4_stdlib_double/1048576|436406|412255|5.53%|446367|465273|-4.24%|
|aos_to_soa3_stdlib_float/192|19.7|13.1|33.50%|19.7|10.9|44.67%|
|aos_to_soa3_stdlib_float/3072|317|213|32.81%|315|158|49.84%|
|aos_to_soa3_stdlib_float/24576|3866|2372|38.64%|3194|2257|29.34%|
|aos_to_soa3_stdlib_float/786432|142402|112994|20.65%|135272|111304|17.72%|
|aos_to_soa3_stdlib_double/192|23|23|0.00%|23|19.1|16.96%|
|aos_to_soa3_stdlib_double/3072|590|430|27.12%|586|691|-17.92%|
|aos_to_soa3_stdlib_double/24576|6944|5812|16.30%|7090|7684|-8.38%|
|aos_to_soa3_stdlib_double/786432|278704|245164|12.03%|283166|299725|-5.85%|
||||||||
|soa_to_aos4_stdlib_float/256|26.3|19.7|25.10%|26.3|26.3|0.00%|
|soa_to_aos4_stdlib_float/4096|486|319|34.36%|420|420|0.00%|
|soa_to_aos4_stdlib_float/32768|3706|3429|7.47%|4169|3803|8.78%|
|soa_to_aos4_stdlib_float/1048576|164774|156301|5.14%|164821|160779|2.45%|
|soa_to_aos4_stdlib_double/256|45.9|32.8|28.54%|45.9|26.3|42.70%|
|soa_to_aos4_stdlib_double/4096|737|565|23.34%|736|493|33.02%|
|soa_to_aos4_stdlib_double/32768|9021|8731|3.21%|8944|8797|1.64%|
|soa_to_aos4_stdlib_double/1048576|407176|416744|-2.35%|412842|413210|-0.09%|
|soa_to_aos3_stdlib_float/192|26.2|16.4|37.40%|26.2|16.4|37.40%|
|soa_to_aos3_stdlib_float/3072|422|267|36.73%|422|270|36.02%|
|soa_to_aos3_stdlib_float/24576|3664|2264|38.21%|3435|2256|34.32%|
|soa_to_aos3_stdlib_float/786432|134980|116856|13.43%|129989|116928|10.05%|
|soa_to_aos3_stdlib_double/192|29.6|23|22.30%|29.6|18.5|37.50%|
|soa_to_aos3_stdlib_double/3072|483|453|6.21%|476|599|-25.84%|
|soa_to_aos3_stdlib_double/24576|6131|6054|1.26%|5839|6853|-17.37%|
|soa_to_aos3_stdlib_double/786432|250600|250158|0.18%|244182|283454|-16.08%|

### CODE REVIEW

File: benchmarks/README.md [hunk 1]
```diff
@@ -14,9 +14,16 @@ For some features it also makes sense to compare with alternative implementation
...
 
-You can use CMake options ``BENCHMARKS_ISPC_TARGETS`` and ``BENCHMARKS_ISPC_FLAGS`` to set specific target or ISPC compilation switches. For example, ``-DBENCHMARKS_ISPC_TARGETS=avx512skx-i32x8,avx2-i32x8 -DBENCHMARKS_ISPC_FLAGS="-O3 --woff"``.
+You can use CMake options ``BENCHMARKS_ISPC_TARGETS`` and ``BENCHMARKS_ISPC_FLAGS`` to set specific target or ISPC compilation switches. For example, ``-DBENCHMARKS_ISPC_TARGETS=avx512skx-i32x8,avx2-i32x8 -DBENCHMARKS_ISPC_FLAGS="-O3 --woff"``. Note that using auto-dispatch (i.e. specifying more than one target on the command line) will noticeable affect short benchmark runs.
 
-To run benchmarks, you need to execute them individually. They will be located in `benchmarks` folder of your install location.
+To run benchmarks, you need to execute them individually. They will be located in `benchmarks` folder of your install location. You can also run `make test` to verify that benchmarks are built correctly and execute successfully.
+
+To do a performance measurements, it is recommended to do a frequency stabilization on your system. This is OS and hardware specific and there no universal recipe, but here are the things that might work:
```
Comment (aneshlya) [inline benchmarks/README.md:21]: To do a performance measurements, it is recommended to do a frequency stabilization on your system->
To do performance measurements, it is recommended to do frequency stabilization on your system

File: benchmarks/README.md [hunk 2]
```diff
@@ -14,9 +14,16 @@ For some features it also makes sense to compare with alternative implementation
...
+1. Disable CPU frequency scaling on Linux, as suggested in Google Benchmark [documentation](https://github.com/google/benchmark/blob/main/docs/user_guide.md#disabling-cpu-frequency-scaling)
+2. Control the CPU core where the benchmark is running. This is especially important for heterogeneous CPUs. On Linux you can use `taskset` to pin the process to specific core. For examples:
```
Comment (aneshlya) [inline benchmarks/README.md:23]: For examples: -> For example

**Review by aneshlya — CHANGES_REQUESTED** [review]
LGTM!
Only minor changes in the doc are needed.

File: benchmarks/README.md [hunk 3]
```diff
@@ -14,9 +14,16 @@ For some features it also makes sense to compare with alternative implementation
 
 Use ``-DISPC_INCLUDE_BENCHMARKS=ON`` in your CMake configuration flags to enable building benchmarks. If ``--recurse-submodules`` was not used during repo cloning, CMake will fetch [Google Benchmark](https://github.com/google/benchmark) submodule. By default ISPC built from sources will used for building benchmarks. If you need to use pre-build ISPC binary, you can turn off building ISPC by setting ``-DISPC_BUILD=OFF`` and provide ISPC in your `PATH` environment variable.
 
-You can use CMake options ``BENCHMARKS_ISPC_TARGETS`` and ``BENCHMARKS_ISPC_FLAGS`` to set specific target or ISPC compilation switches. For example, ``-DBENCHMARKS_ISPC_TARGETS=avx512skx-i32x8,avx2-i32x8 -DBENCHMARKS_ISPC_FLAGS="-O3 --woff"``.
+You can use CMake options ``BENCHMARKS_ISPC_TARGETS`` and ``BENCHMARKS_ISPC_FLAGS`` to set specific target or ISPC compilation switches. For example, ``-DBENCHMARKS_ISPC_TARGETS=avx512skx-i32x8,avx2-i32x8 -DBENCHMARKS_ISPC_FLAGS="-O3 --woff"``. Note that using auto-dispatch (i.e. specifying more than one target on the command line) will noticeable affect short benchmark runs.
```
[the remark this answers is not in the document]
Comment (dbabokin) [reply benchmarks/README.md]: fixed

File: benchmarks/README.md (hunk 1 above)
Comment (dbabokin) [reply benchmarks/README.md]: fixed

File: benchmarks/README.md [hunk 4]
```diff
@@ -14,9 +14,16 @@ For some features it also makes sense to compare with alternative implementation
...
 
-To run benchmarks, you need to execute them individually. They will be located in `benchmarks` folder of your install location.
+To run benchmarks, you need to execute them individually. They will be located in `benchmarks` folder of your install location. You can also run `make test` to verify that benchmarks are built correctly and execute successfully.
```
[the remark this answers is not in the document]
Comment (dbabokin) [reply benchmarks/README.md]: fixed

File: benchmarks/README.md (hunk 2 above)
Comment (dbabokin) [reply benchmarks/README.md]: fixed

**Comment by dbabokin** [comment]
Docs are fixed.

**Review by aneshlya — APPROVED** [review]

**Review by JeffRous — APPROVED** [review]
Looks good! Thanks for running the perf tests. Might need to do some work on the soa_to_aos4 doubles implementation but can be later.

**Pull request merged**